clients/drcachesim/tests/burst_reattach.cpp

```diff
@@ -0,0 +1,158 @@
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+/* This application links in drmemtrace_static and acquires a trace during
+ * a "burst" of execution and memory allocations (malloc() and free())
+ * in the middle of the application. It then detaches. Later it re-attaches
+ * and detaches again for several times.
+ */
+
+/* Like burst_static we deliberately do not include configure.h here */
+
+#include "dr_api.h"
+#include "drmemtrace/drmemtrace.h"
+#include <assert.h>
+#include <iostream>
+#include <fstream>
+#include <string>
+#include <math.h>
+#include <stdlib.h>
+
+bool
+my_setenv(const char *var, const char *value)
+{
+#ifdef UNIX
+    return setenv(var, value, 1 /*override*/) == 0;
+#else
+    return SetEnvironmentVariable(var, value) == TRUE;
+#endif
+}
+
+// Test recording large values that require two entries.
+ptr_uint_t
+return_big_value(int arg)
+{
+    return (((ptr_uint_t)1 << (8 * sizeof(ptr_uint_t) - 1)) - 1) | arg;
+}
+
+static int
+do_some_work(int arg)
+{
+    static const int iters = 100;
+    double *val = new double; // libc malloc is called inside new
+    *val = arg;
+    for (int i = 0; i < iters; ++i) {
+        *val += *val + (double)return_big_value(i);
+    }
+    double temp = *val;
+    delete val; // libc free is called inside delete
+    return (temp > 0);
+}
+
+static void
+exit_cb(void *)
+{
+    const char *funclist_path;
+    drmemtrace_status_t res = drmemtrace_get_funclist_path(&funclist_path);
+    assert(res == DRMEMTRACE_SUCCESS);
+    std::ifstream stream(funclist_path);
+    assert(stream.good());
+    std::string line;
+    bool found_malloc = false;
+    bool found_return_big_value = false;
+    while (std::getline(stream, line)) {
+        assert(line.find('!') != std::string::npos);
+        if (line.find("!return_big_value") != std::string::npos)
+            found_return_big_value = true;
+        if (line.find("!malloc") != std::string::npos)
+            found_malloc = true;
+    }
+    assert(found_malloc);
+    assert(found_return_big_value);
+}
+
+int
+main(int argc, const char *argv[])
+{
+    if (!my_setenv("DYNAMORIO_OPTIONS",
+                   "-stderr_mask 0xc"
+                   " -client_lib '#;;-offline -record_heap"
+                   // Test the low-overhead-wrapping option.
+                   " -record_replace_retaddr"
+                   // Test large values that require two entries.
+                   " -record_function \"malloc|1&return_big_value|1\"'"))
+        std::cerr << "failed to set env var!\n";
+
+    for (int i = 0; i < 100; i++) {
```

Comment:
The do_some_work() is shrunk in this one and there are some tweaks in the options (i.e., removal of rstats_to_stderr).  In the latest commit I also removed more options and did more tweaks (removed record_heap and use and recording of return_big_value()). All these help reduce the overhead a bit (27%). If we are ok with the extra overhead we could just increase the reattach count in burst_malloc, or tweak the burst_malloc to reduce its overhead (e.g., simplify its do_some_work() ) . I also thought that this test might evolve and get disabled in some settings, so it might be good to keep it separate.

Reply:
If you removed all the function recording, it no longer reproduces the bug?

Reply:
Yes